### PR TITLE
[KOGITO-721] - Introducing Jobs Service

### PR DESCRIPTION
**Pull request by ricardozanini**
See:
https://issues.redhat.com/browse/KOGITO-721

In this PR we introduce a new CRD to handle [Jobs Service](url) feature.

-------------8<--------------
TODO:

- [x] Handle defer function in the controller
- [x] Handle CR Status
- [x] Write unit tests for controller
- [x] Test in an actual cluster
- [x] Add Integration to `KogitoApps` (Job Service Route must be added to them)
- [x] CLI integration
- [x] Infinispan integration via `KogitoInfra`
- [x] Write documentation

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

**Comment by ricardozanini** [comment]
Depends on: https://github.com/RHsyseng/operator-utils/pull/27

**Comment by ricardozanini** [comment]
@radtriste I'm still working on it and I have a couple things to do before releasing for review, but it's functional. If you want to give it a go, I'll be back on Jan 7th. Happy new year!

**Comment by radtriste** [comment]
First test seems ok. Hope to have smoke tests soon integrated so I can implement one for this case ;)

**Comment by ricardozanini** [comment]
@radtriste CLI integration is ready, could you please take a look on it?
Plus, please reaply the CRD for Data Index and Jobs Service since I changed the `ServiceURI`spec to just `URI` for Infinispan integration. Should be done for Kafka as well on https://issues.redhat.com/browse/KOGITO-614

**Comment by ricardozanini** [comment]
@spolti @radtriste @sutaakar @xiezhang7 could you guys please take a look? I intend to merge it tomorrow, so Friday we might have a pretty cool 0.7.0-rc1 sent to dev-preview channel.

File: README.md [hunk 1]
```diff
@@ -355,6 +362,75 @@ property in your domain data, this data will be reflect automatically in the Dat
...
+ # Deploy to the cluster
+ $ oc create -f deploy/crds/app.kiegroup.org_v1alpha1_kogitojobsservice_cr.yaml -n my-project
+```
+
+### Enabling Persistence with Infinispan
+
+Jobs Service supports persistence with Infinispan by setting the property `spec.infinispan.useKogitoInfra` to `true` in the
+CR or the flag `--enable-persistence` in the CLI.
+
+When doing this Kogito Operator deploys a new Infinispan server via Infinispan Operator for you within the same namespace
```
Comment (sutaakar) [inline README.md]: Does job service use same Infinispan server as Data index, or different server is used?

File: cmd/kogito/command/shared/converters_test.go [hunk 2]
```diff
@@ -23,25 +23,51 @@ import (
 
 func Test_fromStringToImage(t *testing.T) {
```
Comment (sutaakar) [inline cmd/kogito/command/shared/converters_test.go]:
```suggestion
func TestFromStringToImageStream(t *testing.T) {
```

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+                    properties:
+                      configMapKeyRef:
+                        description: Selects a key of a ConfigMap.
+                        properties:
+                          key:
+                            description: The key to select.
+                            type: string
+                          name:
+                            description: 'Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+                              TODO: Add other useful fields. apiVersion, kind, uid?'
```
Comment (sutaakar) [inline deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: Is it needed or can be removed?

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+                      secretKeyRef:
+                        description: Selects a key of a secret in the pod's namespace
+                        properties:
+                          key:
+                            description: The key of the secret to select from.  Must
+                              be a valid secret key.
+                            type: string
+                          name:
+                            description: 'Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+                              TODO: Add other useful fields. apiVersion, kind, uid?'
```
Comment (sutaakar) [inline deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: Same here.

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+                authRealm:
+                  description: Name of the Infinispan authentication realm. This sets
+                    the property infinispan.client.hotrod.auth_realm.
+                  type: string
+                credentials:
+                  description: SecretCredentialsType is the data structure for specifying
+                    credentials within a Secret
+                  properties:
+                    passwordKey:
+                      description: PasswordKey is the password for the user.
```
Comment (sutaakar) [inline deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: Isn't it rather a key pointing to value in Secret holding password value?

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+                      type: string
+                    secretName:
+                      description: SecretName is the name of the secret where the
+                        credentials are set
+                      type: string
+                    usernameKey:
+                      description: UsernameKey is the user name for the user who will
```
Comment (sutaakar) [inline deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: Same as above.

File: README.md (hunk 1 above)
Comment (ricardozanini) [reply README.md]: Same. If there's a `KogitoInfra` deployed in the namespace, Jobs Service will use that.

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml (hunk 3 above)
Comment (ricardozanini) [reply deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: Generator issues, I'll remove it.

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml (hunk 5 above)
Comment (ricardozanini) [reply deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: Yes, indeed.

File: README.md (hunk 1 above)
Comment (sutaakar) [reply README.md]: ok

**Comment by radtriste** [comment]
Tests ran:

- CR installation has no problem
  - without persistence OK
  - with persistence OK
- CLI installation
  - without persistence (kogito install jobs-service) OK
  - with persistence (kogito install jobs-service --enable-persistence) **KO**
    Infinispan operator is installed but there is no infinispan instance created ...

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml (hunk 3 above)
Comment (ricardozanini) [reply deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: @sutaakar this is generated by the Kubernetes API, we can't do much about it. It's from the `EnvVar` resource.

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml (hunk 4 above)
Comment (ricardozanini) [reply deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: Again, this is generated by the Kubernetes resource `ResourceRequirements`, there's nothing we can do. :(

**Comment by ricardozanini** [comment]
@radtriste could you do another test run? Thanks!

File: deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml (hunk 3 above)
Comment (sutaakar) [reply deploy/crds/app.kiegroup.org_kogitojobsservices_crd.yaml]: ok

**Comment by radtriste** [comment]
CLI problem is corrected

File: README.md [hunk 7]
```diff
@@ -355,6 +362,75 @@ property in your domain data, this data will be reflect automatically in the Dat
...
+There's some options to customize the Jobs Service deployment with CLI. 
+Run `kogito install jobs-service --help` to understand and set them according to your requirements.  
+
+#### Installing the Kogito Jobs Service with the Operator Catalog (OLM)
+
+If you are running on OpenShift 4.x, you can use the OperatorHub user interface to create the Kogito Jobs Service resource. 
+In the OpenShift Web Console, go to **Installed Operators** -> **Kogito Operator** -> **Kogito Jobs Service**. 
+Click **Create Kogito Jobs Service** and create a new resource as shown in the following example:
+
+```ỳaml
```
Comment (spolti) [inline README.md]: small typo `y

File: README.md (hunk 1 above)
Comment (spolti) [inline README.md]: s/via/using?

File: README.md [hunk 8]
```diff
@@ -355,6 +362,75 @@ property in your domain data, this data will be reflect automatically in the Dat
...
+### Enabling Persistence with Infinispan
+
+Jobs Service supports persistence with Infinispan by setting the property `spec.infinispan.useKogitoInfra` to `true` in the
+CR or the flag `--enable-persistence` in the CLI.
+
+When doing this Kogito Operator deploys a new Infinispan server via Infinispan Operator for you within the same namespace
+that you're deploying Jobs Service. It also sets all information regarding server authentication.
+
+For this to work, bear in mind that Infinispan Operator must be installed in the namespace. If the Kogito Operator was installed
+via OLM, changes are that the Infinispan Operator would also be installed. If installed manually, you will also have to manually
```
Comment (spolti) [inline README.md]: the "changes are that the Infinispan Operator would also be installed" is not clear to me. maybe rephrase?

File: cmd/kogito/command/shared/install_operator.go [hunk 9]
```diff
@@ -60,6 +61,11 @@ func SilentlyInstallOperatorIfNotExists(namespace string, operatorImage string,
 func MustInstallOperatorIfNotExists(namespace string, operatorImage string, cli *client.Client, silence bool) (installed bool, err error) {
 	log := context.GetDefaultLogger()
 
+	if util.GetBoolOSEnv("DEBUG") {
```
Comment (spolti) [inline cmd/kogito/command/shared/install_operator.go]: not sure if I followed this, what is the purpose to skip the operator installation if the debug env is set?

Comment (ricardozanini) [reply cmd/kogito/command/shared/install_operator.go]: When we are debugging the CLI using `operator-sdk` feature it's not interesting to have the CLI to install the operator into the namespace since we're running it locally. We might change the env var, though. Like `SKIP_OPERATOR` or `LOCAL`. WDYT?

File: README.md (hunk 7 above)
Comment (ricardozanini) [reply README.md]: What an eye! I thought it was a dust in my screen.

File: cmd/kogito/command/shared/install_operator.go (hunk 9 above)
Comment (spolti) [reply cmd/kogito/command/shared/install_operator.go]: I think SKIP_OPERATOR would be better.

**Comment by ricardozanini** [comment]
@spolti just reviewed the typos. Many thanks!
